### ISSUE TITLE
Add support to free a processor group

### ISSUE THREAD

**Issue by gabrieltanase42** [StanfordLegion/legion#849]
In our usage of realm we need to to often create and destroy processor groups. Think a database where for a query request I create a processor group, execute query in that processor group and then destroy the processor group. The processor group sizes are arbitrarily and depend on the query needs and the load on other cores.

**Comment by gabrieltanase42** [comment]
I already have the fix for this feature but I don;t know how to contribute it back. could you please help me with some quick instructions on how I can push my branch with changes to your repository ?

I patched on top of master branch;

**Comment by gabrieltanase42** [comment]
[free_group_diff.txt](https://github.com/StanfordLegion/legion/files/4663342/free_group_diff.txt)

**Comment by gabrieltanase42** [comment]
I was able to fork , clone and push my diff. I hope I did the pull request the proper way :)
Pls let me know if there are any issues cherry picking my commit

**Comment by streichler** [comment]
I have created a `free_group` fork in our gitlab repo here: https://gitlab.com/StanfordLegion/legion/-/tree/free_group

My first tweak to the changes was to make the API a little more consistent:
1. Create a `ProcessorGroup` handle that's a subclass of `Processor` (just like `UserEvent` is a subclass of `Event`)
2. Changed the creation method from `Processor::create_group` to `ProcessorGroup::create_group`, leaving the old one as a deprecated wrapper.
3. Changed the proposed `Processor::free_group` method to `ProcessorGroup::destroy`.

Some more changes are needed behind the scenes to deal with deferred destruction and the like, but this should be it for externally-visible changes.  @gabrieltanase42 can you let me know if these API tweaks are going to cause major pain?

@lightsighter I tweaked the Legion code to use the non-deprecated creation method, but did not try to preserve the `ProcessorGroup` type, which will be needed if you want to destroy the groups later.  Would you like me to take a shot at that, or is it easier for you to make those changes?

**Comment by lightsighter** [comment]
> Would you like me to take a shot at that, or is it easier for you to make those changes?

@streichler How are you proposing to change the mapper interface to destroy processor groups?

**Comment by streichler** [comment]
>  How are you proposing to change the mapper interface to destroy processor groups?

I wasn't really.  :)  If I poked at it, it'd just be to remember the groups as `ProcessorGroup` instead of `Processor` so that they can all be cleaned up at shutdown time.

**Comment by lightsighter** [comment]
Is that really necessary or is that just nice to have? Is there some massive overhead with `ProcessorGroup` that would cause mappers to want to delete the groups that they create after they are done using them?

**Comment by gabrieltanase42** [comment]
Hi Mike,
for us the usage scenario is like this: we have a server running and this
server receives requests to run database queries. Now for each query
received we create an IO processor group and a compute processor group, we
execute the query and then we destroy the processor group when query is
over. Depending on the query we create different processor groups with
different amounts of threads/cores. We have a scheduler that essentially
decides hot to manage the io threads and the compute cores and others like
memory and database transactions, etc.

For the patch that I proposed do you all see anything that is wrong or you
you are thinking to make it more general like adding a termination event ?


Thx,
--Gabriel

**Comment by streichler** [comment]
@gabrieltanase42 I have the additional changes in this branch and passing CI, so all that it's waiting on to be merged into the master branch is a confirmation from you that the API changes are acceptable.

**Comment by lightsighter** [comment]
@gabrieltanase42 I don't really care so much about changes to the Realm API. I'm mostly questioning @streichler about how he sees the Legion APIs above this changing since processor groups aren't explicitly exposed in Legion.

**Comment by gabrieltanase42** [comment]
I looked at the diff and I see a more comprehensive restructuring of create group in addition to destroy. The only aspect I care about is to make sure we don't leak resources when destroying. 
With this in mind I would suggest to extend your test/realm/proc_group.cc with one more unit test where we do something like this:

```
for (int i=0;i<1000000;++i){
  ProcessorGroup pgrp = ProcessorGroup::create_group(all_cpus);
  pgrp.destroy(Event::merge_events(pgrp_events));
 // assert something here like no processors groups leaked if possible; personally I was just watching with top the memory consumption and making sure it doesn't go up during the test.  I also used some jemalloc functionality for leaks;
}
```
The specific syntax is not much of a concern. It seems that right now you made it into a method of the processor group which is fine with me.


Thx for your help

**Comment by lightsighter** [comment]
I think I would also like to understand how expensive it is to "leak" a processor group? How many resources are being leaked, and is there overhead in the scheduling of a normal processor if it becomes a part of too many groups?

**Comment by streichler** [comment]
A ProcessorGroup probably isn't more than a few hundred bytes when its queue is empty, but a processor must check the queue for each group of which it is a member every time it schedules a task.  Checking a queue requires taking a probably-uncontended lock and then just a load or two, so I doubt you'd notice anything until you've got 100's of groups sitting around.

**Comment by streichler** [comment]
>Is it possible that we also accept NO_EVENT and then we just terminate the group without any waiting (user guarantees there are no more tasks in flight) ???

Yes, `NO_EVENT` is always a legal value to pass as a precondition, as long as you've ensured that there really are no remaining preconditions at the time you call the `destroy` method.

**Comment by streichler** [comment]
> We also need to be absolutely sure we don't leak resources when destroying.
With this in mind I would suggest to extend your test/realm/proc_group.cc with one more unit test where we do something like this:

There are some quirks of how Realm allocates and reuses memory that make it a bit hard to detect leaks inside of a Realm test, so I went with an "inductive" proof that we're not leaking by verifying the IDs are reused:
https://gitlab.com/StanfordLegion/legion/-/blob/free_group/test/realm/proc_group.cc#L266-273

**Comment by gabrieltanase42** [comment]
The interface then looks good for us. I also see in your test calling destroy with no argument which is how we will be using it.

**Comment by streichler** [comment]
This is merged into master with commit bcc1c62.

**Comment by lightsighter** [comment]
Updated Legion to use the new processor group API with  888fa4744b. Legion also cleans up processor groups correctly now at the end of a run. Legion still does not currently support deletion of processor groups in the middle of a run, waiting for someone to have a demonstrable need for it.
